docs/enhancements/agent-upgrade.md

```diff
@@ -0,0 +1,189 @@
+
+### Non-Goals
+
+Hosts will still need to be rebooted with a new version of the ISO if there
+are changes in the part of the agent that performs the initial registration.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster creator I want to be able to generate an ISO, boot and register
+hosts with a version of the service, then upgrade the service to a version that
+requires a newer version of the agent, and finally install the cluster without
+having to reboot the hosts.
+
+#### Story 2
+
+As a cluster Creator I want to be able to generate an ISO, boot and register
+hosts with a version of the service, then downgrade the service to a version
+that requires an older version of the agent, and finally install the cluster
+without having to reboot the hosts.
+
+### Implementation Details/Notes/Constraints
+
+A new `compatible-agent` validation will be added to the service to check if the
+agent is compatible.
+
+The logic for that validation will be a simple string comparison of the value
+of the `AGENT_DOCKER_IMAGE` environment variable with the value of the
+`discovery_agent_version` stored in the `hosts` table.
+
+To make this comparison useful it is important that the `AGENT_DOCKER_IMAGE`
+environment variable passed to the service doesn't use the `:latest` tag, as
+that renders the comparison useless. In the _SaaS_ stage environment that
+variable is already using a tag, so that shouldn't be a problem:
+
+```
+$ oc get deployment assisted-service -n assisted-installer-stage -o json | \
+jq '.spec.template.spec.containers[].env[] | select(.name == "AGENT_DOCKER_IMAGE")'
+{
+  "name": "AGENT_DOCKER_IMAGE",
+  "value": "registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141"
+}
+```
+
+In the integration and development environments we are using the `:latest` tag,
+which effectively disables this feature.
+
+The description of the `compatible-agent` validation will be the following:
+
+> The installation cannot start just yet because this host's agent is in the
+> process of being upgraded to a newer version, please wait, this could take
+> a few minutes.
+
+A new `upgrade-agent` step will be added to the service and to the agent.
+
+When the service receives a request for next steps from a host it will check if
+that host is in one of the `preparing-for-installation`, `installing` or
+`installing-in-progress` states. In it is then the service will return the next
+steps as usual. If it isn't then the service will retrieve from the database the
+result of the `compatible-agent` validation. If the validation passed then the
+service will return the next steps as usual, otherwise it will replace all the
+usual steps the `upgrade-agent` step.
+
+When the agent receives the `agent-upgrade` step it will run the equivalent of
+`systemctl restart agent`. That will restart the agent service, which will
```

Comment:
We remove agent image on service restart and it will require re-pulling old image and then pulling new one. I assume they will have many common labels but we already saw systems where it took more then 3 minutes to pull image. Maybe we should consider retry policy on [RunAgent](https://github.com/openshift/assisted-installer-agent/blob/89405454e4f080235a8dd06374bab49be1424231/src/agent/agent.go#L40) itself? I mean if next_step_runner got exit_code "upgrade_required" we should not restart agent, we should only restart register process. It will allow us to pull only one image and mostly will not set host to disconnected as we will be back quicker than 3 minutes

Reply:
Done.